feat: allow for overriding of busybox image

congiv opened the pull request:
Hi!

I'm trying to use this chart in an environment where we have a requirement to only use an internally hosted image registry.

I've added values overrides for the `busybox` image that's being pulled from Dockerhub by default to allow for that.

mealingr commented on chart/templates/iq-server-deployment.yaml:
```diff
@@ -221,7 +221,7 @@ spec:
       initContainers:
         - name: {{ .Release.Name }}-set-iq-persistence-ownership
-          image: busybox:1.28
+          image: {{ include "nexus-iq-server-ha.busyboxImage" . }}
```
Thanks for this suggestion.

There's a couple of other places where we use busybox which I think we'll also need to change:

https://github.com/sonatype/nexus-iq-server-ha/blob/53cfbb3daddd82bb5a24dd82c178b0c052da0797/chart/templates/delete-old-aggregate-logs-cronjob.yaml#L21

https://github.com/sonatype/nexus-iq-server-ha/blob/450c35fc9eb182d37a69dffa1dd204e68f3dea6b/chart/values.yaml#L630

mealingr commented on chart/templates/_helpers.tpl:
```diff
@@ -5,3 +5,7 @@
+
+{{- define "nexus-iq-server-ha.busyboxImage" -}}
+{{- if .Values.busybox.imageRegistry }}{{ .Values.busybox.imageRegistry }}/{{ .Values.busybox.image }}:{{ .Values.busybox.tag }}{{- else }}{{ .Values.busybox.image }}:{{ .Values.busybox.tag }}{{- end }}
+{{- end -}}
```
(newline at the end just for codestyle)
```suggestion
{{- end -}}

```

mealingr commented on chart/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
 apiVersion: v2
 name: nexus-iq-server-ha
-version: 171.0.0
+version: 171.1.0
```
Shouldn't need to change the version so I'd make sure this is up-to-date with main

mealingr commented:
You may also need to change your fluentd forwarder/aggregator images https://artifacthub.io/packages/helm/bitnami/fluentd/5.5.12 (only if these are being used though)

mealingr commented:
I think it would be helpful to update https://github.com/sonatype/nexus-iq-server-ha/blob/main/chart/README.md#L709 as well.

I've opened a separate PR with these comments addressed if it makes merging easier
https://github.com/sonatype/nexus-iq-server-ha/pull/26

congiv commented:
@mealingr thanks for all of the suggestions!

> I've opened a separate PR with these comments addressed if it makes merging easier

I was going through your comments one by one and just saw this one last, whoops! However you want to handle the merge is fine by me, I'll keep updating this PR unless you tell me you'd prefer it be handled on #26.

> to "{{ tpl (include \"nexus-iq-server-ha.busyboxImage\" .) $}}"

I think I've addressed all of your suggestions except this one, as this causes `helm template` and `helm lint` to fail. It appears to be an issue with scope on my branch, as when I change the `nexus-iq-server-ha.busyboxImage` to use something that is defined under the `fluentd` overrides, it templates correctly, but for the scope of the fluentd subchart. For example:

```yaml
# _helpers.tpl
{{- define "nexus-iq-server-ha.busyboxImage" -}}
{{ .Values.forwarder.configFile }}
{{- end -}}
```
renders:
```yaml
 initContainers:
   image: 'fluentd.yaml'
```


I see on #26 [there are additional parenthesis around `(.Values.busybox)`](https://github.com/sonatype/nexus-iq-server-ha/pull/26/commits/93bbb556b213bff49b0ae4bca1fb2c31650f6dc8#diff-a695820e1f93e7decd902e54573160d357bc57f3faeec7e501cc7d2c607baaaaR10) which allows the template and lint to run successfully, but the content of the `nexus-iq-server-ha/charts/fluentd/templates/aggregator-statefulset.yaml` still appears to be wrong. This is what I get when I run `helm template chart` from the source branch of #26:

![image](https://github.com/sonatype/nexus-iq-server-ha/assets/47261237/8f9ce60f-8ed0-487f-b0a5-0c16c0dbea29)

congiv commented:
My best guess right now is that even though the templates are shared between the parent chart and the subcharts, the values of the parent are inaccessible from the subchart.

From: https://helm.sh/docs/chart_template_guide/subcharts_and_globals/
> 1. A subchart is considered "stand-alone", which means a subchart can never explicitly depend on its parent chart.
> 2. For that reason, a subchart cannot access the values of its parent.

mealingr commented:
> If you see this issue too and agree that this explains it, an alternative suggestion might be to just let the user override `fluentd.aggregator.initContainers` directly. Not ideal that they'd have to define the busybox image twice, but it works.

Thanks for the spot, I think you're right.

This seems to be a long-standing issue in helm
https://github.com/helm/helm/issues/6699
https://github.com/helm/helm/pull/6876
https://github.com/helm/helm/issues/12323

Still I think we can get around it by making it global (thanks for the link)

I've made the change in https://github.com/sonatype/nexus-iq-server-ha/pull/26 and it seems to work
```
helm template chart >out.txt

# Source: nexus-iq-server-ha/charts/fluentd/templates/aggregator-statefulset.yaml
...
      initContainers:
        - command:
          - sh
          - -c
          - chown -R 1000:1000 /opt/bitnami/fluentd/logs/nexus-iq-server
          image: 'busybox:1.28'
...
```
```
helm template --set global.busybox.image=busybox2 chart >out.txt

# Source: nexus-iq-server-ha/charts/fluentd/templates/aggregator-statefulset.yaml
...
      initContainers:
        - command:
          - sh
          - -c
          - chown -R 1000:1000 /opt/bitnami/fluentd/logs/nexus-iq-server
          image: 'busybox2:1.28'
...
```
unit tests are passing as well via `helm unittest --helm3 chart`

mealingr commented:
I've merged https://github.com/sonatype/nexus-iq-server-ha/pull/26 so I think we can close this one, thanks @congiv

congiv commented:
Thanks!
